crates/biome_css_parser/src/syntax/property/border.rs

```diff
@@ -0,0 +1,130 @@
+use crate::parser::CssParser;
+use crate::syntax::css_dimension::{is_at_length_dimension, parse_regular_dimension};
+use crate::syntax::{is_at_color, parse_color, parse_regular_identifier};
+use biome_css_syntax::{
+    CssSyntaxKind::{self, *},
+    T,
+};
+use biome_parser::diagnostic::expect_one_of;
+use biome_parser::prelude::ParsedSyntax::{Absent, Present};
+use biome_parser::prelude::*;
+
+use super::parse_property_value_with_fallbacks;
+
+/// https://drafts.csswg.org/css-backgrounds/#propdef-border
+///
+/// ```ebnf
+///  border =
+///      <line-width>  ||
+///      <line-style>  ||
+///      <color>       
+///  
+///  <line-width> =
+///      <length [0,∞]>  |
+///      thin            |
+///      medium          |
+///      thick           
+///  
+///  <line-style> =
+///      none    |
+///      hidden  |
+///      dotted  |
+///      dashed  |
+///      solid   |
+///      double  |
+///      groove  |
+///      ridge   |
+///      inset   |
+///      outset
+/// ```
+#[inline]
+pub(crate) fn parse_border_property(p: &mut CssParser) -> ParsedSyntax {
+    // Assumes the parent has confirmed we're at the `border` identifier.
+    let m = p.start();
+    parse_regular_identifier(p).ok();
+    p.expect(T![:]);
+
+    parse_property_value_with_fallbacks(p, parse_css_border).ok();
+
+    Present(m.complete(p, CSS_BORDER_PROPERTY))
+}
+
+/// Parse a complete border shorthand value:
+///     <line-width>  ||
+///     <line-style>  ||
+///     <color>     
+fn parse_css_border(p: &mut CssParser) -> ParsedSyntax {
+    let m = p.start();
+    let mut map = [false; 3];
+    let mut any = false;
+
+    loop {
+        if !map[0] && is_at_line_width(p) {
+            parse_any_line_width(p).ok();
+            map[0] = true;
+            any = true;
+        } else if !map[1] && is_at_line_style(p) {
+            parse_line_style(p).ok();
+            map[1] = true;
+            any = true;
+        } else if !map[2] && is_at_color(p) {
+            parse_color(p).ok();
+            map[2] = true;
+            any = true;
+        } else {
+            break;
+        }
+    }
+
+    if !any {
+        p.error(expect_one_of(
+            &["line width", "line style", "color"],
+            p.cur_range(),
+        ));
+        m.abandon(p);
+        return Absent;
+    }
```

Comment:
I'd really like to wrap this up into a nice macro. it could look like:

```rust
parse_unordered_some! {
  "line width" => is_at_line_width(p) => parse_any_line_width(p),
  "line style" => is_at_line_style(p) => parse_any_line_style(p),
  "color" => is_at_color(p) => parse_any_color(p)
};
```
And would generate the exact code I've selected here. But for the meantime, rewriting this loop isn't terrible.

`parse_unordered_all` would also assert that _all_ of the possible branches are filled, to match the `&&` combinator.

Reply:
Or we can implement a trait like we have for lists, I'm thinking about an error recovery.

Reply:
A trait sounds good too! Especially for recovery. I'm definitely a little out of my current depth in that regard, so if you have any ideas I'd be very happy to hear and talk about them to see what can be the best approach.